MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpress.c

```diff
@@ -172,15 +184,17 @@ EnumerateNvmeDevNamespace (
     Device->BlockIo.WriteBlocks = NvmeBlockIoWriteBlocks;
     Device->BlockIo.FlushBlocks = NvmeBlockIoFlushBlocks;
 
-    //
-    // Create BlockIo2 Protocol instance
-    //
-    Device->BlockIo2.Media         = &Device->Media;
-    Device->BlockIo2.Reset         = NvmeBlockIoResetEx;
-    Device->BlockIo2.ReadBlocksEx  = NvmeBlockIoReadBlocksEx;
-    Device->BlockIo2.WriteBlocksEx = NvmeBlockIoWriteBlocksEx;
-    Device->BlockIo2.FlushBlocksEx = NvmeBlockIoFlushBlocksEx;
-    InitializeListHead (&Device->AsyncQueue);
+    if (Private->NSQA > 1 || Private->NCQA > 1) {

```

Comment:
Is it necessary to check NCQA > 1?

Will NCQA < NSQA ever be true?

Reply:
Assuming NSQA=1 and NCQA=2 is possible, would we want to install BlockIo2? @apop5

Reply:
Further, what about NSQA=2 and NCQA=1? That's not a case I've seen but one I know is possible.

Reply:
I can change the check to && instead of || so there are at least 2 of each before installing BlockIo2, I'm just unsure of the expected behavior of the above cases. 
| # of CQ   | # of SQ    | Behavior |
| ------------- | ------------- | ----------------|
1 | 1 | Blocking Data queue only. Don't install BlockIo2
1 | 2+ | I can see a scenario where this uses both blocking and unblocking - ?
2 +| 1 | I Assume blocking only?
2+ | 2+ | Blocking and unblocking data queue. Install BlockIo2

Reply:
even though it was mentioned that submission/completion queues do not have to match, for our purposes in firmware, we should treat them as needing to match.

I don't think it is worth codifying mis matched queues, I think it is safe to have the following:
Admin Submission/Completion - Always required
Submission/Completion queues 1 - Required if we want to Support BlockIo. If there is not at least 1 submission/completion queue available, then we dont have a way to communicate to the drive. 
Submission/Completion queues 2 - Install BlockIo and BlockIo2

If you want to collapse NSQA and NCQA into one variable, that can work  as well. It would tell you if BlockIo2 is supported.